### PR TITLE
feat(runfiles): Added type hints to `@rules_python//python/runfiles`

### PR DESCRIPTION
**Pull request by UebelAndre**
This change adds mypy types directly to the `rules_python.python.runfiles` library as well as regression testing to CI.

### CODE REVIEW

File: python/runfiles/runfiles.py
Comment (UebelAndre) [inline python/runfiles/runfiles.py:26]: These must be defined earlier to be able to use them as types

File: python/runfiles/runfiles.py
Comment (UebelAndre) [inline python/runfiles/runfiles.py:89]: Same as https://github.com/bazelbuild/rules_python/pull/1654/files#r1435762546

File: python/runfiles/runfiles.py
Comment (aignas) [reply python/runfiles/runfiles.py:26]: nit: I thought we could just use a string for the type annotation and I would have assumed that this would have worked irrespectively of where in the file they are written.

File: python/runfiles/runfiles.py
Comment (aignas) [inline python/runfiles/runfiles.py:119]: Related to my comment above, does the following work if the `_ManifestBased` is below this line?
```suggestion
    def __init__(self, strategy: Union["_ManifestBased", "_DirectoryBased"]) -> None:
```

File: .github/workflows/mypy.yaml
Comment (aignas) [inline .github/workflows/mypy.yaml:25]: it would be great to test against Python `3.8`, which does not have the new style annotations to ensure backwards compatibility with previous Python versions. WDYT?

File: python/runfiles/runfiles.py
Comment (UebelAndre) [reply python/runfiles/runfiles.py:26]: Isn’t it simpler to reorder them in the long run? I agree it’s annoyingly noisy but I think the outcome will benefit future maintainers and contributors.

File: python/runfiles/runfiles.py
Comment (UebelAndre) [reply python/runfiles/runfiles.py:119]: https://github.com/bazelbuild/rules_python/pull/1654#discussion_r1435943418

If this is a nit I’d push back since I feel this is the right thing to do. If this blocks merging then I can reluctantly try to put this back. Just let me know what the position is

File: .github/workflows/mypy.yaml
Comment (UebelAndre) [reply .github/workflows/mypy.yaml]: Yeah makes sense. I'll move this back to 3.8

File: python/runfiles/runfiles.py
Comment (aignas) [reply python/runfiles/runfiles.py:119]: Having to move the code around just to add type annotations is surprising and for ease of review it would be good to split such a change into two - one to add type annotations (but maybe use strings in places as a workaround for Python). That said, I am OK to review this as one PR, I just need more time to do it :)

If you think that the end goal should be the place where we don't use strings, I don't have any objections to this.

Comment (UebelAndre) [reply python/runfiles/runfiles.py:119]: I think the end goal is not to use strings. If the review seems too much then I can try to reduce splash damage but it's still mostly code moves and then tweaks to function signatures. Let me know though!

File: python/runfiles/runfiles.py
Comment (aignas) [inline python/runfiles/runfiles.py:193]: This `if` statement should always be `True` and it would be better to add an assert there. On line `180` there is
`if not remainder or (source_repo, target_repo) not in self._repo_mapping`. This means that if there is no remainder, then the code will return there, but if there is, then the `(None, target_repo) not in some_dict` should be always true because I would expect that value to never be present.

File: tests/runfiles/runfiles_test.py
Comment (UebelAndre) [reply tests/runfiles/runfiles_test.py:141]: I originally had this but it mypy would immediately complain since it doesn't seem to understand the API. I could do this and then use a `cast` but I felt this was the lesser evil. I don't really care which it is though so with that said, which would you prefer? or is there a better path?

File: python/runfiles/runfiles.py
Comment (UebelAndre) [reply python/runfiles/runfiles.py:193]: I can promote this to an assertion

File: tests/runfiles/runfiles_test.py
Comment (aignas) [reply tests/runfiles/runfiles_test.py:141]: SGTM to leave the `assert r is not None` but it may be nice to add a comment as to why we are not using the `unittest` API.

File: python/runfiles/runfiles.py
Comment (aignas) [inline python/runfiles/runfiles.py:190]: consider adding an explanation to the `assert`. If this assert fails, having no message may require the user to read the code. Maybe something like the following could work.

```suggestion
        assert source_repo is not None, f"BUG: if the `source_repo` is None, we should never go past the `if` statement above"
```

**Review by aignas — COMMENTED** [review]
For me this looks good, but I'd like to wait for @rickeylev to have a look at this since this would constrain the `runfiles` library to Python `3.7` and I don't know if this is something we want to do now.

**Comment by UebelAndre** [comment]
For what it’s worth, 3.7 was chosen because it’s EOL and I thought it would be a sufficiently low bar.

File: tests/runfiles/runfiles_test.py
Comment (UebelAndre) [reply tests/runfiles/runfiles_test.py]: Done!

**Review by aignas — APPROVED** [review]
LGTM, thanks

**Pull request merged**
